Resolves #5225: Implemented "Itemized Request Report"

seabeeberry opened the pull request:
Resolves #5225 

Implemented the Itemized Request Report, modelled after the Itemized Distribution Report. When you click on Reports, the option for Requests - Itemized apears at the bottom (this can be placed elsewhere if wanted). This shows the list of requested items and how many are on hand at the moment. This can be filtered by time range.

I implented tests as well to verify intended behavior. I checked for Linting with `bundle exec erb_lint --lint-all` and ran all tests (`bundle exec rake`).

### Type of change

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

Reports -> Requests - Itemized

You should see a list of the requested items, their units, the number of requested items and the number of items on hand. Items which are below the set `onhand_minimum` parameter should be displayed as red in the table.

### Screenshot
<img width="1300" height="667" alt="Screenshot 2025-08-03 at 16 05 44" src="https://github.com/user-attachments/assets/70bf3e03-22e0-4d8f-bfcc-4510bcfbb573" />

cielf reviewed (changes requested):
Hey @seabeeberry -- thank you for taking this on!  It looks pretty good functionally, but I have one note:  

Because the total requested included requests that have already been fulfilled,   making the total on hand red if it's less than the total requested will highlight items that they aren't actually short on.      Using the same criteria as for distribution for making the number red should work, though.

When we've got that addressed, I'll pass it on for technical comments.

seabeeberry commented:
Got it! Will make the necessary changes :)

cielf commented:
Hey @seabeeberry -- It's looking good 

There's one more thing I'd like to tweak:  

I know we said "model it on the distributions-itemized" report -- it turns out there is an outstanding issue on that report, and I'm wondering if we could avoid it for this one, instead of having to make another issue to address it.

If an item is inactive,   the total on hand currently shows "Unknown".  (See Adult Briefs (Large/X-Large)). We made a change so that inactive items always have 0 items a few months ago,  so we no longer need to show "Unknown" -- we can just show the actual level.

Could you make that adjustment?

seabeeberry commented:
I changed the fallback in the view from "Unknown" to 0, hopefully that is the intended fix here. I had added a fallback to the fetch method in my first commit, so that the number on hand is always set to 0 in case it's nil, but that crashed one of the automated tests.

cielf commented:
@seabeeberry   

Hmmm --  This is a case where requests are going to be different than distributions -- because if there has been a distribution,  there has to have been inventory at some point --   and that is simply not the case with requests.    So we might have items that have never had inventory in the list.  

Your solution will give the desired result  in that case,  but I'd like  @dorner's  take on whether it's the best way to handle it -- and it's time to get his overall technical advice anyway.

dorner commented on app/services/request_itemized_breakdown_service.rb:
```diff
@@ -0,0 +1,101 @@
+      .includes(:partner, :organization, :item_requests)
+      .where(id: @request_ids)
+      .flat_map do |request|
+        request.request_items.map do |json_item|
```
Wait... are we using item\_requests _and_ request\_items? Why both?

My understanding is that ItemRequest is the correct approach since it's using actual DB records.

dorner commented on spec/services/request_itemized_breakdown_service_spec.rb:
```diff
@@ -0,0 +1,62 @@
+  before do
+    allow_any_instance_of(View::Inventory).to receive(:quantity_for).with(item_id: item_a.id).and_return(3)
+    allow_any_instance_of(View::Inventory).to receive(:quantity_for).with(item_id: item_b.id).and_return(20)
+    allow_any_instance_of(View::Inventory).to receive(:all_items).and_return([
```
I'd much prefer creating actual inventory using the `create_inventory` method.

seabeeberry commented:
I implemented the rest of the suggestions. Let me know if it needs any changes still!

cielf reviewed (approved):
LGTM.  On to merging!

seabeeberry commented:
@DasTapan sure! should I write you an email or do you have a discord?
